### PR TITLE
[ONNX] [Relay] Dynamic squeeze

### PR DESCRIPTION
**Pull request by CircleSpin**
@AndrewZhaoLuo @anwang2009 @mbrookhart

### CODE REVIEW

File: python/tvm/relay/frontend/onnx.py [hunk 1]
```diff
@@ -2806,7 +2820,9 @@ def _impl_v12(cls, inputs, attr, params):
         alpha = _op.const(attr.get("alpha", 1.0), dtype)
         zero = _op.const(0, dtype)
         one = _op.const(1, dtype)
-        return _op.maximum(zero, x) + _op.minimum(zero, alpha * (_op.exp(x / alpha) - one))
+        out = _op.maximum(zero, x) + _op.minimum(zero, alpha * (_op.exp(x / alpha) - one))
+        print(out)
```
Comment (mbrookhart) [inline python/tvm/relay/frontend/onnx.py:2824]: remove the print

**Review by mbrookhart — COMMENTED** [review]
It looks like there are a couple of formatting/lint issues, and there's one print still in here, but otherwise LGTM

File: python/tvm/relay/frontend/onnx.py [hunk 2]
```diff
@@ -1482,6 +1482,20 @@ def _impl_v12(cls, inputs, attr, params):
             result = _op.expand_dims(result, axis)
         return result
 
+class Squeeze(OnnxOpConverter):
+    """Operator converter for Squeeze."""
+
+    @classmethod
+    def _impl_v1(cls, inputs, attr, params):
+        return _op.squeeze(*inputs, axis=attr["axes"])
```
Comment (AndrewZhaoLuo) [inline python/tvm/relay/frontend/onnx.py:1490]: You must handle the case where "axes" is not given: https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Squeeze-1

File: python/tvm/relay/op/dyn/_transform.py [hunk 3]
```diff
@@ -258,3 +259,24 @@ def _sparse_to_dense_shape_func(output_shape, ndim):
...
+        if not_in_axis:
+            out[out_i] = int64(data[i])
+            out_i += 1
+
+    return out
+
+
+@_reg.register_shape_func("dyn.squeeze", [False, True])
+def dynamic_squeeze_shape_func(attrs, inputs, out_ndims):
+    return [_squeeze_shape_func_input_data(inputs[0], inputs[1], out_ndims[0])]
```
Comment (AndrewZhaoLuo) [inline python/tvm/relay/op/dyn/_transform.py:282]: nit: newline at end of file

File: src/relay/op/dyn/tensor/transform.cc [hunk 4]
```diff
@@ -692,6 +692,66 @@ RELAY_REGISTER_OP("dyn.expand_dims")
     .set_attr<FTVMCompute>("FTVMCompute", ExpandDimsCompute)
     .set_attr<TOpPattern>("TOpPattern", kInjective);
 
-}  // namespace dyn
-}  // namespace relay
-}  // namespace tvm
+bool DynSqueezeRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
+                const TypeReporter& reporter) {
+  ICHECK_EQ(types.size(), 3);
```
Comment (AndrewZhaoLuo) [inline src/relay/op/dyn/tensor/transform.cc:697]: nit: add a comment about each of the types
e.g. [input tensor, axes tensor, output tensor]

File: src/relay/op/dyn/tensor/transform.cc [hunk 5]
```diff
@@ -692,6 +692,66 @@ RELAY_REGISTER_OP("dyn.expand_dims")
...
+  const auto* data = types[0].as<TensorTypeNode>();
+  if (data == nullptr) {
+    return false;
+  }
+  const auto* axes = types[1].as<TensorTypeNode>();
+  if (axes == nullptr) {
+    return false;
+  }
+  ICHECK_EQ(axes->shape.size(), 1);
```
Comment (AndrewZhaoLuo) [inline src/relay/op/dyn/tensor/transform.cc:706]: nit: add error messages to these ICHECKs

e.g. ICHECK_EQ(...) << "Error message here got " << variable << " expected 1".

File: src/relay/op/dyn/tensor/transform.cc [hunk 6]
```diff
@@ -692,6 +692,66 @@ RELAY_REGISTER_OP("dyn.expand_dims")
...
+
+
+Array<te::Tensor> SqueezeCompute(const Attrs& attrs, const Array<te::Tensor>& inputs,
+                                 const Type& out_type) {
+  const auto* out_ttype = out_type.as<TensorTypeNode>();
+  ICHECK(out_ttype != nullptr);
+  Array<IndexExpr> newshape;
+  for (auto val : out_ttype->shape) {
+    if (val->IsInstance<tir::AnyNode>()) {
+      newshape.push_back(val.as<tir::AnyNode>()->ToVar());
```
Comment (AndrewZhaoLuo) [inline src/relay/op/dyn/tensor/transform.cc:722]: Do we need these IF checks? They are always going to be AnyNodes() right?

File: src/relay/op/dyn/tensor/transform.cc [hunk 7]
```diff
@@ -692,6 +692,66 @@ RELAY_REGISTER_OP("dyn.expand_dims")
...
+Expr MakeDynSqueeze(Expr data, Expr axes) {
+  auto attrs = make_object<SqueezeAttrs>();
+  static const Op& op = Op::Get("dyn.squeeze");
+  return Call(op, {data, axes}, Attrs(attrs), {});
+}
+
+TVM_REGISTER_GLOBAL("relay.op.dyn._make.squeeze").set_body_typed(MakeDynSqueeze);
+
+RELAY_REGISTER_OP("dyn.squeeze")
+    .describe(R"code(Squeeze the input tensor at the dimensions given by axes
```
Comment (AndrewZhaoLuo) [inline src/relay/op/dyn/tensor/transform.cc:739]: nit: might want to define what it means to "squeeze". E.g. remove axes (with dimension value 1)

File: src/relay/op/dyn/tensor/transform.cc [hunk 8]
```diff
@@ -692,6 +692,66 @@ RELAY_REGISTER_OP("dyn.expand_dims")
...
+    .set_attrs_type<SqueezeAttrs>()
+    .add_argument("data", "Tensor", "The input tensor.")
+    .add_argument("axes", "Tensor", "The axes to squeeze.")
+    .set_support_level(3)
+    .add_type_rel("DynSqueeze", DynSqueezeRel)
+    .set_attr<FTVMCompute>("FTVMCompute", SqueezeCompute)
+    .set_attr<TOpPattern>("TOpPattern", kInjective)
+    .set_attr<TReshapeOp>("TReshapeOp", true);
+
+}
```
Comment (AndrewZhaoLuo) [inline src/relay/op/dyn/tensor/transform.cc:755]: Cap off namespaces
e.g.
} // namespace dyn
} // namespace ...

File: tests/python/relay/dyn/test_dynamic_op_level3.py [hunk 9]
```diff
@@ -92,6 +92,25 @@ def verify_reshape(shape, newshape, oshape):
...
 
+def test_squeeze():
+    def verify_squeeze(shape, dtype, axis):
+        x = relay.var("x", relay.TensorType(shape, dtype))
+        assert axis is not None
+        np_axis = tuple(axis)
+        axis = relay.var("axis", relay.TensorType([len(axis)], "int64"))
+        squeeze = relay.squeeze(x, axis=axis)
+
+        print(squeeze)
```
Comment (AndrewZhaoLuo) [inline tests/python/relay/dyn/test_dynamic_op_level3.py:103]: Remove prints

**Review by AndrewZhaoLuo — COMMENTED** [review]
LGTM mostly. A few comments

**Comment by AndrewZhaoLuo** [comment]
Please also change the title of this PR to something like "[ONNX] [Relay] Dynamic Squeeze"

**Review by AndrewZhaoLuo — COMMENTED** [review]
Thanks for the changes, LGTM

**Review by mbrookhart — APPROVED** [review]

**Pull request merged**